types/src/transaction/encrypted_payload.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction::{TransactionExecutable, TransactionExecutableRef, TransactionExtraConfig};
+use anyhow::{bail, Result};
+use aptos_crypto::HashValue;
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+
+pub type CipherText = Vec<u8>;
+pub type EvalProof = Vec<u8>;
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub enum EncryptedPayload {
+    V1(EncryptedPayloadV1),
+}
+
+impl Deref for EncryptedPayload {
+    type Target = EncryptedPayloadV1;
+
+    fn deref(&self) -> &Self::Target {
+        let Self::V1(payload) = self;
+        payload
+    }
+}
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub enum EncryptedPayloadV1 {
+    Encrypted {
+        ciphertext: CipherText,
+        extra_config: TransactionExtraConfig,
+        payload_hash: HashValue,
+    },
+    FailedDecryption {
+        ciphertext: CipherText,
+        extra_config: TransactionExtraConfig,
+        payload_hash: HashValue,
+
+        eval_proof: EvalProof,
+    },
+    Decrypted {
+        ciphertext: CipherText,
+        extra_config: TransactionExtraConfig,
+        payload_hash: HashValue,
+        eval_proof: EvalProof,
+
+        // decrypted things
+        executable: TransactionExecutable,
+        decryption_nonce: u64,
```

Comment:
Just a heads up. The crypto lib doesn't have the nonce. @rex1fernando needs to support it later.